sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -383,6 +383,24 @@ trait FileSourceScanLike extends DataSourceScanExec {
     }).flatMap(DataSourceStrategy.translateFilter(_, supportNestedPredicatePushdown))
   }
 
+  @transient
+  protected lazy val pushedDownFilters: Seq[Filter] = translatePushedDownFilters(dataFilters)
+
+  @transient
+  protected lazy val dynamicallyPushedDownFilters: Seq[Filter] = {
+    if (dataFilters.exists(_.exists(_.isInstanceOf[execution.ScalarSubquery]))) {
+      // Replace scalar subquery to literal so that `DataSourceStrategy.translateFilter` can
+      // support translate it. The subquery must has been materialized since SparkPlan always
+      // execute subquery first.
+      val normalized = dataFilters.map(_.transform {
```

Comment:
should `selectedPartitions` also use the normalized data filters?

Reply:
I feel a bit of inconsistency here. The partition filter can also contain scala subquery, when is it normalized?

Reply:
Seems `selectedPartitions` does not accept subquery expression, it filter out the dpp filter first.

https://github.com/apache/spark/blob/e326b9b6565b1e2b697d41714060e2ff9e424514/sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala#L252-L257

Then in `dynamicallySelectedPartitions`,  normalized the partition filter.

https://github.com/apache/spark/blob/e326b9b6565b1e2b697d41714060e2ff9e424514/sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala#L268-L281

Reply:
For scalar subquery, I think we should turn it into literal so that it can be used to prune partitions?

Reply:
The `outputPartitioning` and `outputOrdering` use `selectedPartitions` when it is bucket scan which is before do execution. Besides, `selectedPartitions` is a public field, I'm afraid it would break something if we assume it only be called at runtime...